tests/cases/compiler/circularReferenceInReturnType.ts

```diff
@@ -0,0 +1,11 @@
+// @strict: true
+// @noEmit: true
+
+declare function fn1<T>(cb: () => T): string;
+const res1 = fn1(() => res1);
+
+declare function fn2<T>(): (cb: () => any) => (a: T) => void;
+const res2 = fn2()(() => res2);
+
+declare function fn3<T>(): <T2>(cb: (arg: T2) => any) => (a: T) => void;
+const res3 = fn3()(() => res3);
```

Comment:
The only diff between those 2 is the `T2` type parameter. The `fn2` case works today but the `fn3` doesn't - even though its return type doesn't actually depend on `T2`.

Reply:
How does fn3 fail? It looks like inference still fails, given that `res3: (a: unknown) => void` -- but I'm not sure, and I don't see how to observe what's inferred for T2. Maybe the return type of fn3 should be `(a: T) => T2` in the test?

Reply:
> How does fn3 fail?

We end up with: "'res3' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)"

> It looks like inference still fails, given that res3: (a: unknown) => void

It doesn't exactly fail, it's just that there are no inference candidates here so the constraint is used and that's `unknown`. This isn't meant to be used as a compelling use case for this PR, just a super minimal repro. The better motivating example is showcased in `circularReferenceInReturnType2.ts`.

> I don't see how to observe what's inferred for T2.

It's just `unknown`. I re-checked this with `console.log(typeToString(inference.typeParameter), typeToString(inference.inferredType))` before exiting `getInferredType`.

> Maybe the return type of fn3 should be (a: T) => T2 in the test?

That's not good here because that reintroduces the circularity because the return type starts to depend on a type parameter again.